lib/erb_lint/parser.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module ERBLint
+  # Contains the logic for generating the file tree structure used by linters.
+  module Parser
+    END_MARKER_NAME = 'erb_lint_end_marker'
+
+    class << self
+      def parse(file_content)
+        require 'nokogiri'
+        require 'htmlentities'
+
+        file_content += "<#{END_MARKER_NAME}>"\
+          'This is used to calculate the line number of the last line'\
+          "</#{END_MARKER_NAME}>"
+
+        file_content_with_erb_tags = file_content.gsub(/<%(.+?)%>/m) do |_match|
+          "<erb>#{HTMLEntities.new.encode($1)}</erb>"
+        end
+
+        file_content_with_erb_tags = escape_erb_tags_in_strings(file_content_with_erb_tags)
+
+        file_tree = Nokogiri::XML.fragment(file_content_with_erb_tags)
+        validate_tree(file_tree)
+
+        file_tree
+      end
+
+      def filter_erb_nodes(node)
+        node.search('*').select do |element|
+          element.name != 'erb' && element.name != END_MARKER_NAME
+        end
+      end
+
+      private
+
+      def escape_erb_tags_in_strings(file_content)
+        scanner = StringScanner.new(file_content)
+
+        while scanner.skip_until(/'|"/)
+          open_string_type = scanner.matched
+
+          string_start = scanner.pos - 1
+          if scanner.skip_until(/#{open_string_type}/).nil?
```

Comment:
this is one of the parts that needs more work. I'm going to rewrite the string validation part by tracking when strings are being opened and closing.

It will also break on things like `<div class=" ' "></div>`